### PR TITLE
Accounts in Database

### PR DESCRIPTION
**Pull request by Alderon86**
## Description
This is a combination of an old Pr "F-HealthyAccountDB" (can be found on my repo) and Neskk's approach.
Added "on-the-fly" support for csv files. No instance restarting is needed, just change your .csv.

Parses given .csv file into database to use the defined number of workers via -w to start scanning.
Accounts are locked to instances.
Csv Parser got rewritten to also add the account level (usefull for l30's and above)
Removed couple of commands like -u and -p because its defined via csv.
Accounts removed from the csv also will get removed from db.
TODO's are marked properly.
## Motivation and Context

coming soon
## How Has This Been Tested?

locale instance
## Screenshots (if appropriate):
coming soon
## Types of changes

- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

thanks @neskk for inspiration

### CODE REVIEW

**Comment by peterbaumert** [comment]
Just a quick thought. If I have seen it right the Accounts table will be dropped on -cd. I think it would make sense not to drop it. :)

**Comment by Alderon86** [comment]
@peterbaumert Thats why its marked as [WIP] :D
But you are right, that shouldnt happen

File: pogom/apiRequests.py [hunk 1]
```diff
@@ -67,6 +68,7 @@ def parse_remote_config(account, api_response):
 
     remote_config = api_response['responses']['DOWNLOAD_REMOTE_CONFIG_VERSION']
     if remote_config.result == 0:
+        Accounts.set_tempban(account)
```
Comment (friscoMad) [inline pogom/apiRequests.py]: Please handle this during the login as it is the only point it makes sense and we should avoid messing module concerns.

File: pogom/models.py [hunk 2]
```diff
@@ -110,6 +110,358 @@ def get_all(cls):
...
+                'shadowban': account.get('shadowban', None),
+                'warning': account.get('warning', None),
+                'banned': account.get('banned', None)}
+
+    # TODO: Add high_lvl_accounts support.
+    @staticmethod
+    def get_accounts_in_use(min_level=1, max_level=40):
+        query = (Accounts
+                 .select()
+                 .where((Accounts.instance_name == args.status_name) &
```
Comment (friscoMad) [inline pogom/models.py]: pass the status_name as a parameter we should avoid using globals

File: pogom/models.py [hunk 3]
```diff
@@ -110,6 +110,358 @@ def get_all(cls):
...
+            if len(line) == 0 or line.startswith('#'):
+                continue
+
+            fields = line.split(",")
+            fields = map(str.strip, fields)
+            if num_fields < 0:
+                num_fields = len(fields)
+
+            if num_fields == 2:
+                account = {
```
Comment (friscoMad) [inline pogom/models.py]: I am quite sure the can be rewritten shomehow to avoid having to declare defaults twice.

File: pogom/schedulers.py [hunk 4]
```diff
@@ -650,10 +650,9 @@ def empty_queues(self):
 
     # How long to delay since last action
     def delay(self, last_scan_date):
-        return max(
-            ((last_scan_date - datetime.utcnow()).total_seconds() +
-             self.args.scan_delay),
-            2)
+        return min(max(((last_scan_date - datetime.utcnow()).total_seconds() +
```
Comment (friscoMad) [inline pogom/schedulers.py]: ?? This seem totally unrelated.

File: pogom/search.py [hunk 5]
```diff
@@ -351,8 +350,8 @@ def search_overseer_thread(args, new_location_queue, control_flags, heartb,
     they can be tried again later, but must wait a bit before doing do so
     to prevent accounts from being cycled through too quickly.
     '''
-    for i, account in enumerate(args.accounts):
-        account_queue.put(account)
+    add_accounts_to_queue(args, account_queue, db_updates_queue, args.workers,
+                          max_level=29, init=True)
```
Comment (friscoMad) [inline pogom/search.py]: I don't think this will happen anytime soon but sometime ago there was people scanning with lv30 maybe a flag should be added, maybe we can add it when it is needed.

File: pogom/models.py [hunk 6]
```diff
@@ -110,6 +110,358 @@ def get_all(cls):
...
+         .where((WorkerStatus.username == account['username']))
+         .execute())
+
+    # Sets the ban flag of an account. Re-use impossible.
+    @staticmethod
+    def set_permban(account):
+        (Accounts(username=account['username'],
+                  in_use=False,
+                  instance_name=None,
+                  tempban=True)
```
Comment (friscoMad) [inline pogom/models.py]: permban

File: config/config.ini.example [hunk 7]
```diff
@@ -152,7 +152,7 @@
 #wh-types:[]                    # List of events to be sent: pokemon, gym, raid, egg, tth, gym-info, pokestop, lure, captcha. (default= nothing)
 #wh-threads:                    # Number of webhook threads; increase if the webhook queue falls behind. (default=1)
 #wh-retries:                    # Number of times to retry sending webhook data on failure (default=5)
-#wh-timeout:                    # Timeout (in seconds) for webhook requests (default=1).
+#wh-timeout:                    # Timeout (in seconds) for webhook requests (default=2).
```
Comment (sebastienvercammen) [inline config/config.ini.example]: Default has been 1 for a while, and the new webhook updates have changed timeout config items. Could you rebase to latest so it's easier to keep track?

File: pogom/models.py [hunk 8]
```diff
@@ -110,6 +110,391 @@ def get_all(cls):
         return results
 
 
+# TODO: Add support for high_lvl_accounts.
+class Accounts(LatLongModel):
```
Comment (sebastienvercammen) [inline pogom/models.py]: DB model names should be singular (Account).

File: pogom/models.py [hunk 9]
```diff
@@ -110,6 +110,391 @@ def get_all(cls):
...
+                 .order_by(Accounts.level.asc(),
+                           Accounts.last_modified.asc())
+                 .dicts())
+        return query
+
+    @staticmethod
+    def get_accounts(number, min_level=1, max_level=40, init=False):
+        query = []
+        accounts = []
+        # Try to re-use previous accounts for this instance
```
Comment (sebastienvercammen) [inline pogom/models.py]: Dot `.`

File: pogom/search.py [hunk 10]
```diff
@@ -1327,3 +1342,38 @@ def is_paused(control_flags):
...
+
+
+def add_accounts_to_queue(args, account_queue, db_updates_queue, number,
+                          min_level=1, max_level=40, init=False):
+    accounts = []
+    retry_delay = 0
+    loaded_accounts = 0
+    while len(accounts) < number:
+        if len(accounts):
+            # if we've looped through here once, then we have loaded accounts.
```
Comment (sebastienvercammen) [inline pogom/search.py]: Capitalize

> If

File: pogom/models.py [hunk 11]
```diff
@@ -110,6 +110,391 @@ def get_all(cls):
...
+    # TODO: Add high_lvl_accounts support.
+    @staticmethod
+    def get_accounts_in_use(min_level=1, max_level=40):
+        query = (Accounts
+                 .select()
+                 .where((Accounts.instance_name == args.status_name) &
+                        (Accounts.fail == 0) &
+                        (Accounts.in_use == 1) &
+                        (Accounts.shadowban == 0) &
+                        (Accounts.banned == 0) &
```
Comment (sebastienvercammen) [inline pogom/models.py]: On selects, we never seem to use the field `tempban`. If it's because we assume a tempban will be unbanned, perhaps we should add a minimum amount of time before the account is re-used? Otherwise it'll keep getting selected to use in scans.

Same problem with shadowbans: a shadowban will eventually be lifted. How/when is the account re-enabled if it's no longer shadowbanned?

For both, the simplest solution is to add a field that tracks since when they were tempbanned/shadowbanned (e.g. `shadowban_since`), and to include accounts in selects if that field is > x hours/days.

File: pogom/models.py [hunk 12]
```diff
@@ -110,6 +110,391 @@ def get_all(cls):
...
+                         .where((Accounts.instance_name == args.status_name) &
+                                ~(Accounts.username << usernames))
+                         .execute())
+
+        log.debug('Got {} accounts.'.format(len(accounts)))
+
+        return accounts
+
+    @staticmethod
+    # Load Accounts and return a list which we are parsing into the db.
```
Comment (sebastienvercammen) [inline pogom/models.py]: > accounts

> DB

File: pogom/models.py (hunk 3 above)
Comment (sebastienvercammen) [reply pogom/models.py]: @friscoMad is correct. Define the defaults for the `account` object outside of the conditional, then update the object according to the values you have available.

e.g.:
```
account = { 'something': 'default' }

if num_fields = x:
    account['something_else'] = 1
elif num_fields = y:
    account['something_else'] = 2
```

File: pogom/models.py [hunk 13]
```diff
@@ -110,6 +110,391 @@ def get_all(cls):
...
+            else:
+                log.error((
+                    'File "%s" has an error on line "%d": ' +
+                    'invalid field count, check syntax.'), args.accountcsv,
+                    num)
+                return False
+        return accounts
+
+    @staticmethod
+    def process_accounts(db_updates_queue):
```
Comment (sebastienvercammen) [inline pogom/models.py]: Add a description of what the function does. The name itself isn't very descriptive.

File: pogom/models.py [hunk 14]
```diff
@@ -110,6 +110,391 @@ def get_all(cls):
...
+                        (Accounts.shadowban == 0) &
+                        (Accounts.banned == 0) &
+                        (Accounts.level == 30))
+                 .order_by(Accounts.last_modified.asc())
+                 .limit(10)  # Don't spend too much time checking accounts.
+                 .dicts())
+
+        if not query:
+            log.warning('Found no available high-level accounts in database.')
+            return hlvl_account
```
Comment (sebastienvercammen) [inline pogom/models.py]: For clarity, this should be `return None`.

File: pogom/models.py [hunk 15]
```diff
@@ -110,6 +110,391 @@ def get_all(cls):
...
+        return hlvl_account
+
+    # Updates the DB account after an action to show it's still in use.
+    @staticmethod
+    def heartbeat(account):
+        (Accounts(username=account['username'],
+                  in_use=True,
+                  instance_name=args.status_name,
+                  last_modified=datetime.utcnow())
+         .save())
```
Comment (sebastienvercammen) [inline pogom/models.py]: The problem with these methods is that we're bypassing the DB update queue and we don't group account updates. The performance won't scale well.

File: pogom/models.py [hunk 16]
```diff
@@ -110,6 +110,391 @@ def get_all(cls):
...
+
+    # Resets all instance-flagged accounts to set them free for re-use.
+    @staticmethod
+    def reset_instance(keep_instance_name=False):
+        instance_name = args.status_name if keep_instance_name else None
+        (Accounts.update(in_use=False, instance_name=instance_name)
+                 .where(Accounts.instance_name == args.status_name)
```
Comment (sebastienvercammen) [inline pogom/models.py]: Rather than call it `reset_instance`, something like `release_accounts` could be more descriptive because that's what the query does, except for allowing an instance to keep its `instance_name` but resetting `in_use`.

File: pogom/models.py [hunk 17]
```diff
@@ -110,6 +110,391 @@ def get_all(cls):
...
+                  tempban=True)
+         .save())
+        (WorkerStatus
+         .delete()
+         .where((WorkerStatus.username == account['username']))
+         .execute())
+
+    # Sets the ban flag of an account. Re-use impossible.
+    @staticmethod
+    def set_permban(account):
```
Comment (sebastienvercammen) [inline pogom/models.py]: As explained above, all these update functions should only change the value in memory, flag the account as "dirty" and use the DB update queue to update all the dirty accounts as one group.

File: pogom/models.py [hunk 18]
```diff
@@ -341,6 +725,46 @@ def get_appearances_times_by_spawnpoint(pokemon_id, spawnpoint_id,
 
         return list(itertools.chain(*query))
 
+    @classmethod
+    def get_pokemons_nearby(cls, center, timestamp):
+
+        # Maximum distance, where we see pokemons nearby
```
Comment (sebastienvercammen) [inline pogom/models.py]: > nearby.

File: pogom/models.py [hunk 19]
```diff
@@ -1917,6 +2347,40 @@ def parse_map(args, map_dict, scan_coords, scan_location, db_update_queue,
     ScannedLocation.update_band(scan_location, now_date)
     just_completed = not done_already and scan_location['done']
 
+    # Checking if account is shadowbanned
+    if (nearby_pokemon or wild_pokemon) and not args.no_pokemon:
+
+        # Create list of present pokemon IDs for blinded check
```
Comment (sebastienvercammen) [inline pogom/models.py]: Dots in all the comments here.

> .

File: pogom/models.py [hunk 20]
```diff
@@ -1917,6 +2347,40 @@ def parse_map(args, map_dict, scan_coords, scan_location, db_update_queue,
...
+        for p in nearby_pokemon:
+            nearby_pokemon_ids.append(p.pokemon_id)
+        for p in wild_pokemon:
+            nearby_pokemon_ids.append(p.pokemon_data.pokemon_id)
+
+            # Remove common pokemons from seen
+        rare_finds = [p for p in nearby_pokemon_ids if p not in common_ids]
+    # Checking if found only common pokemons
```
Comment (sebastienvercammen) [inline pogom/models.py]: Indentation.

File: pogom/models.py [hunk 21]
```diff
@@ -1917,6 +2347,40 @@ def parse_map(args, map_dict, scan_coords, scan_location, db_update_queue,
...
+
+        if len(rare_finds) == 0:
+            # Get nearby active pokemons from database
+            status['nonrares'] += 1
+            if status['nonrares'] >= 3:
```
Comment (sebastienvercammen) [inline pogom/models.py]: Is 3 a good enough count? What's the average count of non-rare scans on regular accounts?

File: pogom/models.py [hunk 22]
```diff
@@ -2456,11 +2918,15 @@ def encounter_pokemon(args, pokemon, account, api, account_sets, status,
                               hlvl_account['username'])
 
             if ('ENCOUNTER' in enc_responses and
-                    enc_responses['ENCOUNTER'].status != 1):
+                    enc_responses['ENCOUNTER'].status == 0):
```
Comment (sebastienvercammen) [inline pogom/models.py]: Keep `!= 1` for now, until the code below has been added for the other response codes.

File: pogom/models.py [hunk 23]
```diff
@@ -2677,6 +3142,41 @@ def clean_db_loop(args):
...
 
+                # Some quite inactive Accounts in use? Reset them.
+                # Caused by hard shut-down or more workers than needed.
+                # TODO: review account cleanup
+                query = (Accounts
+                         .update(in_use=False)
+                         .where((Accounts.in_use == 1) &
+                                (Accounts.last_modified <
+                                 (datetime.utcnow() - timedelta(minutes=15))))
+                         .execute())
```
Comment (sebastienvercammen) [inline pogom/models.py]: These account cleanups should be in their own methods in `Accounts` and be called by the `clean_db_loop`.

File: pogom/models.py [hunk 24]
```diff
@@ -42,7 +42,7 @@
 flaskDb = FlaskDB()
 cache = TTLCache(maxsize=100, ttl=60 * 5)
 
-db_schema_version = 20
+db_schema_version = 21
```
Comment (fosJoddie) [inline pogom/models.py]: is this correct to increase schema version?

Comment (Alderon86) [reply pogom/models.py]: Once this is merged we dont want to get back to older version to not mess up setups @fosJoddie

**Comment by fosJoddie** [comment]
- it says accounts are kept per instance, but there's no support to have account files separate per instance, in fact it will happily allow you to start two instances with their own accounts file, and proceeds to load/unload each others accounts and fudge things up.

I was under the impression that the logic would load files from instance A and lock them to instance A, and then load files for instance B separate from that? 

The logic will gladly take an account that is level 0 (not done tutorial) over an account that has done tutorial and cause some issues from this (not sure if 100% sure thing to happen or somewhat random). Since accounts last a very short time, it's bad if restart will use lvl 0 accounts before lvl 2.

**Comment by Alderon86** [comment]
@fosJoddie this logic will be applied with #2424 for now, we just inserting accs into the db and use them fpr scanning `assuming` your are only using one! .csv file

File: pogom/models.py (hunk 24 above)
Comment (friscoMad) [reply pogom/models.py]: This should be 22 now

**Comment by friscoMad** [comment]
I have been thinking about what @fosJoddie posted and we will probably break things badly for some people, my use case is naive just 1 instance 1 file with accounts, that should be ok with this PR but I am not that sure that people with multiple instances would understand or be able to have the same setup.
The logic says that one could just put a file with a lot of accounts and just let it go but that's now what people is used to and they don't want to waste a lot of accounts if something goes wrong so they will keep as low as possible for some time.
That could or not be an issue with several instances depending on how/when they change accounts, etc.

With this PR only 1 instance should be in charge of the accounts file but there is no way to comunicate to other instances that A is the instance in charge.

So my proposal would be to add all this code and PR but try to avoid as much user errors as possible, let RM work 2 ways:

- as it is now but with file reload, keep everything in memory (or mark all accounts as coming from this instance and just delete accounts from this instance)
- add a new flag as -cd or script that load and control the accounts file (we can make it that it stops after the first load or keep watching the file)

That way #1 can keep working as we advance down the road if #2 does not fit all use cases, and #2 it is clear enough to avoid issues with several instances, single instances on the other hand will need to run #2 alone to hot change accounts.

If this seems too much work then at least I will ask some people with several instances to run this PR for some time to see if it fits all needs.

And I also came accross a possible problem, maybe it has already been thought if accounts are removed from DB but workers are not forced to abandon those accounts they will be readded when they are changed (in_use to False or whatever) and reused again if there is no error with it, at least it is a bit weird as we will be using an account not in the file.

**Comment by neskk** [comment]
@friscoMad and @sebastienvercammen 
This all started back in August when Alderon asked me to rework some old code from Arengo that had accounts on database and rebase it to work with latest develop branch. There was a lot of code scattered across many files and it was a bit nightmare'ish to get everything working. In this process I dropped the shadowban detection code that was present in the base code so that I could focus on getting accounts working, view commit: https://github.com/RocketMap/RocketMap/pull/2424/commits/99015382cdf6f82746e89e6be98f0e01d5ce9089
After this step I immediately knew there was still a lot of work required to manage how accounts were kept "in sync" with instances, how to correctly rotate accounts and prevent account query updates from being done all over the place. I postponed working on this and until last month I hadn't change anything yet. Because of this, Alderon decided to create this PR/branch based of what I had worked so far and improved the code to make it work.
I had some free time and I really wanted this feature done correctly so in the holidays I restarted work on my branch and since then I already made:
- Account Manager, handles all account queues in the same place.
- Moved database updates from `Account.save()/query()` to db bulk upsert queue.
- I knew we had to keep track of which accounts were being used by which instances and I knew that simply using `status_name` was a mistake since, if undefined it would be assigned with the Process ID which would most likely change in every startup, causing new accounts to be allocated instead of old ones. Created a function that hashes several configuration parameters to derive an instance_id, if you relaunch the scanners, firstly it will automatically try to get accounts that were previously allocated before allocating new accounts if previously used aren't enough (bans, fails, etc).
- I don't think it's useful having accounts in database and in a file... The way I see it account importing/exporting should only be done in specific times and I think the way to go is having a web interface that lets you do this without resorting to files. Maintaining the updates to the CSV file is kinda duplicating the database, it can lead to setting different values in the file and in db potentially creating byzantine errors, where something is going wrong and you can't exactly pinpoint where.
- There's a lot more changed in PR #2424: captchas can be handled in all situations, CSV parser improved, high-level encounters improved, ban detection improved, queue management improved.
I recommended @sebastienvercammen to skip this PR in favor of #2424 because their starting points are pretty much the same and I got account rotation to work correctly using the database.
I'm not saying PR #2424 is absolutely ready and flawless but I think it's already more stable and reliable than this one, for instance, I've been using my PR in my public map, managing an account pool of over 10k accounts. Also I don't get why we're going to merge  #2376 so that then I have to change pretty much everything it changed, the first time I spoke with Seb about this I told him that maybe 2376 merge first was ok, but now I only see duplicate work... again, the starting point of this PR is based on my first commit on #2424 and even rebasing my PR after this one will be messy since they pretty much overlap changes in the same places.
Never the less I really wanted to thank @Alderon86 for his patience and dedication, he's been the one that wants to get things done fast and is always available to get his hands dirty. I didn't come here to say: _hey look, shutdown this PR because it uses code from my branch_ or any of that, I said I was ok with Alderon starting his work based of my branch, but now he's still dealing with issues that I already solved and my PR pushed past this one. No point in having to work more on this if most issues are addressed in my PR.

I just wanted to clarify my view and justify the reason I created PR #2424. If anyone wants to help and review code I'm accepting it gladly 👍

**Comment by sebastienvercammen** [comment]
Per our discussion in Discord in the RM channels and between Foxbabe and myself, we've decided the following:

* This PR will be closed in favor of #2424.
* #2376 separates DB model from behavior changes and was posted a full **two months** before #2424 (so it was also better tested). It was a better candidate to get merged first so behavior changes could be approached separately, which is important for a big breaking change like this.
* We don't appreciate someone pushing their own PR to replace another. If you see an existing PR with an issue, we expect you to contribute to the original PR as teamwork to build something better **together** instead of creating a separate PR that would replace it. Although you say this wasn't your intention, it's the result.
* @neskk was aware of @Alderon86 and they had previously worked together: I was told by both that #2376 is based on earlier work from @neskk. This only gets me to ask why it was impossible for @neskk to approach @Alderon86 and continue improving together over this two-month period.
* After several explanations sent to @neskk over several days that #2376 would be the first to be merged, after which we'd rebase #2424 to focus on the behavior changes, several members on the team (myself included) still continued to receive messages about why #2424 was better and why #2376 should be ignored. All instructions or requests for teamwork were ignored repeatedly.
* #2424 now includes changes that #2376 would need to adopt not to conflict, and changes that are meant as fixes/optimizations, so #2376 is in an older state and would need updating. Since this would also mean duplicate work to rebase and rework #2424, the voted decision is to merge #2424.
* The way in which 2376 vs 2424 was handled is not OK and will never be repeated. In the future, all PRs that are looking to be a direct replacement for another PR by taking the PR's work and adding more on top will automatically be delayed and the original PR will be first to be considered as a merge candidate. (This does not apply to PRs that each have different approaches to solving a problem.)

@neskk has received this explanation in PM on Discord. The response was the following:

> [3:39 PM] neskk: Alderon could have worked with me, his initial push was straight from my branch. It seems you're the one causing a drama as Alderon told me you would... I've always talked with him and got along, and the last time I talked with him he said he was OK doing this. You're making a storm in a glass of water. Nice way of saying thanks to me for my work because it seems really not important to you guys.
> [3:41 PM] neskk: Plus, collaboration doesn't not have it's credits properly accounted for as I told you last week when you merge spawn-spawn rework which had 90% of the code done and debugged by me.
> [3:44 PM] neskk: Not that this stops me from collaborating, but on this instance Alderon could have asked to collaborate in my branch and I would have agreed no problem... he wanted to make his branch based on mine, I said it was fine, no problem whatsover.
I feel you're being unfair to me but Ok.

**This is not kindergarten.** So please stop acting like it. "Drama" is a thing that happens on school grounds, or when children are spreading rumors about who's in a relationship with whom. After having been removed two times already from our Discord channels for behavior, this is the final straw.

Either you do your best to be a decent person and to work together and learn, and you learn to communicate properly with people instead of insulting them each time something is explained to you, or you're not meant to be on a collaboration platform in the first place.

@neskk Should you still want to contribute, I can guarantee that each of your future PRs will be considered objectively and without bias. But any comments you make on Github or Discord will be looked at with renewed scrutiny. This is the third and final time I've had to explain this to you.

**Comment by neskk** [comment]
Sometimes I wonder why I keep spending my time on this project since all you do @sebastienvercammen is hold a grudge towards me... but it's fine, I forgive you :)

>  All instructions or requests for teamwork were ignored repeatedly.

This is not true whatsoever, Alderon and I talked a bit and I tried to help him saying what changes were going to help 2376 get to the same database model I was using in 2424.
It immediately became clear that it would be easier just to go to 2424 and I think Alderon realized that as well. If I've pushed @Alderon86 in anyway into doing something he was against or not comfortable with I'm trully sorry.
I'm also trully sorry for saying something that was said to me in private and telling it to you @sebastienvercammen to be used against me publicly.

> (...) This only gets me to ask why it was impossible for @neskk to approach @Alderon86 and continue improving together over this two-month period.

Maybe I wasn't clear enough... I didn't work on this for the last two months hence not having collaborated with Alderon. Due to my lack of time I'm guessing Alderon was tired of waiting and decided to take action in his own hands - which I encouraged back then. To say that I refused to collaborate is untrue and unfair. Alderon could also have asked me to collaborate on my branch but decided to create his own... I have no problem whatsoever with this but you're painting me as an uncooperative person which is not true at all.

As I've said in Discord directly to Seb (I think he misunderstood my point):
For me this subject can be closed and we can all move forward and focus on what's important: improving RM.

**Comment by sebastienvercammen** [comment]
@neskk 

* Being kicked from our channels twice after group discussions for both cases is hardly what I would call a personal grudge.
* On enough occasions, you were told to work with Alderon on #2376 for the DB model: 5 times on Jan. 6 (you were very persistent in trying to get us to change our decision), once on Jan. 8.
* Your inability to work as a team is because we kept telling you what RM was doing and what we expected: 2376 as DB model, 2424 for behavior changes. Yet you kept up the endless pushing to get your own PR to replace 2376. It's not related to "a lack of time", since your persistence in pushing was active and very noticeable in these past few days.

> I'm also trully sorry for saying something that was said to me in private and telling it to you @sebastienvercammen to be used against me publicly.

I'm not. It's easy to change a person's words to fit your own narrative in private when you believe no one is listening. I make my conversations public so the team has the full context, with no words edited or removed. That includes your attempt at an insult, which you've quickly retracted just now in PM on Discord:

> [5:01 PM] neskk: I should also apologise to you, because what Alderon said was that there was a drama going on and I assumed it coming from you. I may have made a false assumption, so... I'm trully sorry if I offended you in anyway :smiley:
> [5:05 PM] Seb: If you would truly be apologetic, you wouldn't give conflicting information on Discord and on Github. I've replied to your comment and this is my last response to you. Everything is sent to the team for context and the team will continue as they prefer.

As for "misunderstandings":

> As I've said in Discord directly to Seb (I think he misunderstood my point):
> For me this subject can be closed and we can all move forward and focus on what's important: improving RM.

The relevant context from PM:

> [4:22 PM] Seb: And that's OK, there's no need to talk. If you have any other questions, I'm here to answer them.
> [4:23 PM] neskk: You make it look like I'm the bad guy...
> [4:26 PM] Seb: There's a conflict that happened and that has been on-going for several weeks, which included several people on RM's team. It popped up several times in several of our channels. I gathered all feedback and asked for a vote on a decision. A decision was made, I then forwarded it to you. And from there on, everything is as you can see in our last messages.
> [4:26 PM] Seb: That's all that happened.
> [4:27 PM] Seb: It's my task to make sure problems are resolved and don't get repeated. So that's what I did, with Foxbabe and with the team's input.
> [4:27 PM] neskk: I will try to defend myself and say my peace in github but for me this subject is closed and we should move forward.
> [4:28 PM] Seb: It's not up to you to decide how the team decides to move forward with it - people were invited to share their opinions. But the decision has been made yes, and we've explained what we expect and what we offer. If you don't follow it, there won't be a fifth time.
> [4:29 PM] Seb: Everything else, you do what you want. That's up to you.
> [4:31 PM] neskk: fifth?! what the hell?
> [4:32 PM] Seb: My bad, that said "fourth or fifth" before I edited. I must've cut a wrong part. And it refers to the PM + Github message being my third warning to you.

*("The team", "messages" and "decisions" refering to me linking the full context to our team members, as well as the decisions explained above about both PRs.)*

As mentioned over and over again: the full context was shared with the team, you've been informed of the circumstances and you **will not** receive a fourth warning. We've had enough of your games.

We're usually flexible, and we try to accomodate everyone, but you've been so disruptive that we don't have the freedom to extend that flexibility to you. So we won't anymore.

**Comment by neskk** [comment]
> Your inability to work as a team is because we kept telling you what RM was doing and what we expected: 2376 as DB model, 2424 for behavior changes. Yet you kept up the endless pushing to get your own PR to replace 2376. It's not related to "a lack of time", since your persistence in pushing was active and very noticeable in these past few days.

And I kept telling you that it didn't make sense to separate things. I really didn't have a choice in the matter, you decided it was that way and my opinion was not taken in consideration.
When I referred to "lack of time" I was referring to the last two months that 2376 was here and I didn't work on it. I feel you're not respecting me, because yes I have lack of time, have exams and other stuff to do. I did something for my personal use that I thought would be good for the RM community, decided to share it in my PR #2424. There was already #2376 but it was a bit stagnated, not many reviews and it had a lot of things that needed to be addressed. You were asking me to do double work and just because I didn't agree with that idea you stopped answering to my "_pushy_" messages.
Attack my credibility, ban me, kick me out, do whatever you like. I got kicked out of PR because I basically disagreed with you on how we were handling lvl30s and botting. I acknowledge I may have been a bit rude on some occasions and maybe have done some things that were polarizing the devs, but my only crime was contesting your decision and asking for justification. Now you make it sound like I'm a big bully and always cause trouble and that is not accurate.
I say you hold a personal grudge because you always depict me as being as being the bad guy and trouble maker when most of this confusion and drama could have been avoided all together.
I really don't want to spend more time arguing about this, I sincerely apologize to @Alderon86 and to the  _team members_ that are being caught up in this discussion. I did not intend for this to happen.

**Comment by lmao22** [comment]
@sebastienvercammen

Lol... This is not the first time I'm asking myself if you are a worse clone of Adolf Hitler.
The great dictator, development stands still on RM because only you can decide what to do.

This one is dead.

**Comment by sebastienvercammen** [comment]
Removed an insulting comment from a spam account created just now. User @lmao22 has also been blocked from the repository.

To finalize this PR: Both the discussion and the PR are continued in #2424.
